python_modules/dagster/dagster/_core/definitions/op_invocation.py

```diff
@@ -324,6 +326,55 @@ def _resolve_inputs(
     return input_dict
 
 
+def _key_for_result(result: MaterializeResult, context: "BoundOpExecutionContext") -> AssetKey:
+    if result.asset_key:
+        return result.asset_key
+
+    if len(context.assets_def.keys) == 1:
+        return next(iter(context.assets_def.keys))
+
+    raise DagsterInvariantViolationError(
+        "Unable to resolve unset asset_key for MaterializeResult, set explicitly when constructing."
+    )
+
+
+def _handle_gen_event(
+    event: T,
+    op_def: "OpDefinition",
+    context: "BoundOpExecutionContext",
+    output_defs: Mapping[str, OutputDefinition],
+    outputs_seen: Set[str],
+) -> T:
+    if isinstance(
+        event,
+        (AssetMaterialization, AssetObservation, ExpectationResult),
+    ):
+        return event
+    elif isinstance(event, MaterializeResult):
+        asset_key = _key_for_result(event, context)
+        output_name = context.assets_def.get_output_name_for_asset_key(asset_key)
+        outputs_seen.add(output_name)
```

Comment:
per https://github.com/dagster-io/dagster/pull/16913#discussion_r1341657563, this was one of the potential paths i saw, but i didn't feel confident this was the perfect place for the fix - it somehow felt a stop gap to tell the later part (where we do type checking for outputs) to skip check this and i'm not sure if this is the right place to do, cos the root cause of this is that the generator path didn't pick up [the fix](https://github.com/dagster-io/dagster/pull/16883), because all generator return types are "Any" [here](https://github.com/dagster-io/dagster/blob/9ec0f16d9c83ed3d75ee609ba222e30e22ba7170/python_modules/dagster/dagster/_core/definitions/inference.py#L65).

so two questions for you:
- maybe, at minimum, should we rename this to be something more intuitive, such as "outputs_to_skip_type_checking" 😅 ?
- do you think a longer term solution is to not cast MaterializeResult to Nothing and differentiate the handling for MaterializeResult/AssetCheckResult (i.e. events that indicate opt-out of io managers) from the io manager path?

Reply:
> do you think a longer term solution is to not cast MaterializeResult to Nothing and differentiate the handling for MaterializeResult/AssetCheckResult (i.e. events that indicate opt-out of io managers) from the io manager path?

potentially, depends on the nitty gritty details

Reply:
> maybe, at minimum, should we rename this to be something more intuitive, such as "outputs_to_skip_type_checking"

maybe,I think in the current state while "materializations" still map to the low level "output" primitive the current naming makes enough sense

Reply:
> in the current state while "materializations" still map to the low level "output" primitive the current naming makes enough sense

ok this makes sense.